lib/pure/options.nim

```diff
@@ -176,21 +310,239 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+            else:
+              cases.add (ident: nil, check: optionCase[1][2],
+                body: optionCase[2])
+    else:
+      error "Unrecognized structure of cases", optionCase
+  if noneCase == nil:
+    error "Must have a \"none\" case"
+  if elseCase.body == nil and cases.len == 0:
+    error "Must have a \"some\" case"
+  var
+    optionSym = genSym(nskLet)
+    value = nnkDotExpr.newTree(optionSym, newIdentNode("val"))
+  var caseStmt = nnkCaseStmt.newTree(
+    value
+  )
+  for c in cases:
+    var ofBranch = nnkOfBranch.newTree(
+      c.check,
+      newStmtList()
+    )
+    if c.ident != nil:
+      ofBranch[1].add newLetStmt(c.ident, value)
+    ofBranch[1].add c.body
+    caseStmt.add ofBranch
+  if elseCase.body != nil:
+    if elseCase.ident != nil:
+      caseStmt.add nnkElse.newTree(newStmtList(
+        newLetStmt(elseCase.ident, value),
+        elseCase.body
+      ))
+    else:
+      caseStmt.add nnkElse.newTree(elseCase.body)
+  result = quote do:
+    (proc (): auto =
+      let `optionSym` = `option`
+      if `optionSym`.isSome:
+        `caseStmt`
+      else:
+        `noneCase`
+    )()
+  #echo result.repr
+
+proc optCmp*[T](self: Option[T], value: Option[T],
+  cmp: proc (val1, val2: T): bool): Option[T] =
+  ## Comparison operation that return a some iff ``self`` is a some and the
+  ## value of ``cmp`` applied to ``self`` and ``value`` returns true. The
+  ## returned some has the value of ``self``.
+  if self.isSome and value.isSome and cmp(self.val, value.val):
+    return self
+
+template `==`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``==`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 == val2)
+
+template `not`*[T](self: Option[T]): Option[T] =
+  ## Not always returns a ``none`` as a ``none`` can't implicitly get a value
+  none[T]()
+
+template `<=`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``<=`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 <= val2)
+
+template `>=`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``>=`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 >= val2)
+
+template `>`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``>`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 > val2)
+
+template `<`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``<`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 < val2)
+
+proc toOpt*[T](value: Option[T]): Option[T] =
+  ## Procedure with overload to automatically convert something to an option if
+  ## it's not already an option.
+  return value
+
+proc toOpt*[T](value: T): Option[T] =
+  ## Procedure with overload to automatically convert something to an option if
+  ## it's not already an option.
+  some(value)
+
```

Comment:
There is `some` already, isn't this redundant?

Reply:
No, it has two versions, one that take an option and one that doesn't. This means that if the value is an option it will not be wrapped in a second option, but if it's not an option it will be converted into one.

Reply:
@andreaferretti Nim also supports generic programming (including templates), where you may need to support both types with the same code